### PR TITLE
Refactor requests write to `response_stream` once

### PR DESCRIPTION
**Pull request by omus**
Follow up to: https://github.com/JuliaCloud/AWS.jl/pull/516#discussion_r756212793

By making the repeat try/catch the innermost one we can eliminate having to keep track of the number of attempts.

### CODE REVIEW

**Comment by omus** [comment]
I'll call out I've removed the special EOF logic from #516 just to validate it is still needed (I think it is)

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
I'm concerned the tests from #516 don't actually test the EOF behaviour noticed in #515

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
@IanButterworth can you review this? I removed the EOF logic which stops the `response_stream` from being written to on the last failure but the tests you added didn't fail. Do you actually need that behaviour?

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

File: test/issues.jl [hunk 1]
```diff
@@ -190,22 +196,36 @@ try
...
+            io = IOBuffer()
+
+            apply(_incomplete_patch(; data=data, num_attempts_to_fail=4)) do
+                params = Dict("response_stream" => io)
+                @test_throws err_t S3.get_object(bucket, key, params; aws_config=config)
+
+                seekstart(io)
+                retrieved = read(io)
+                @test length(retrieved) == n - 1
+                @test retrieved == data[1:(n - 1)]
```
Comment (omus) [inline test/issues.jl:228]: Note this is exactly the behaviour produced from #516 and is probably not what was intended. The reason this is the case is that for the check `should_write || attempt == max_attempts` the variable `should_write = false` but `attempt = 4` which is equivalent to `max_attempts`.

I personally think we should probably keep the behaviour of having the final attempt always written to `response_stream` even if the result is partial. As an exception is raised already the user has been signalled there's a problem with the data and keeping the partial data could be useful in the case the result is an AWS exception.

**Comment by IanButterworth** [comment]
> I removed the EOF logic which stops the response_stream from being written to on the last failure but the tests you added didn't fail. Do you actually need that behaviour?

I don't see any issue with how this has been re-written. AFAICT, it's a different way of achieving the same behavior: Retry failures aren't passed to the `response_stream` except for the last failed attempt.

The only functional difference I can tell is that on master `response_stream` is written to in all but the type of errors thrown my premature EOF in fixed content length responses. In this PR any error on an attempt stops the buffer from being passed to `response_stream`. 

The reason #516 specifically only prevented the buffer forwarding for #515-related errors was the suggestion somewhere in the reviews (or a slack discussion, I can't remember) that the existing behavior might be needed for some other response type?

If that's not the case, this PR looks good to me

File: test/issues.jl (hunk 1 above)
Comment (IanButterworth) [reply test/issues.jl:228]: That sounds good to me. 

In my experience testing #516 the network intermittent issues causing #515 rarely (perhaps never) happened multiple times in a row, and if that happened such that a attempts are exhausted and an error throws, then passing the incomplete content forward seems fair.

**Comment by omus** [comment]
> The only functional difference I can tell is that on master `response_stream` is written-to in all but the type of errors thrown my premature EOF in fixed content length responses. In this PR _any_ error on an attempt stops the buffer from being passed to `response_stream`.

A quick correction: this PR always writes the buffer from the final attempt to the `response_stream` even if the last attempt was a failure or an incomplete transfer. 

> The reason #516 specifically only prevented the buffer forwarding for #515-related errors was the suggestion somewhere in the reviews (or a slack discussion, I can't remember) that the existing behavior might be needed for some other response type?
> 
> If that's not the case, this PR looks good to me

I'll call out that we probably do want to forward the buffer when errors occur, even EOF errors, as there may relevant information included in that content. The specific example I'm thinking of is an AWS exception where the information is all included in the content so even if we have a incomplete transfer we should pass that information along. Also worth mentioning we already throw an exception when an EOF event occurs so the caller should be adequately notified of the situation.

**Comment by IanButterworth** [comment]
>  this PR always writes the buffer from the final attempt to the response_stream even if the last attempt was a failure or an incomplete transfer.

Yeah, that's what master does too.

**Comment by IanButterworth** [comment]
Ok, I was mistaken.. I thought the `should_write = true` was in the try, so it's state wasn't being reset.

File: test/issues.jl (hunk 1 above)
Comment (IanButterworth) [reply test/issues.jl:228]: The error is propagated, is just whether the incomplete data is also written to the buffer.

**Comment by IanButterworth** [comment]
friendly bump

**Comment by omus** [comment]
I was waiting on a review from @mattBrzezinski but I think we're safe to merge this.

**Comment by omus** [comment]
@mattBrzezinski can you review this? It needs approval. I'll use my admin privileges to merge this at EOD tomorrow otherwise

**Comment by omus** [comment]
It's important to get this merged soon as otherwise in [some cases AWS exceptions will be suppressed](https://github.com/JuliaCloud/AWS.jl/pull/522#issuecomment-980044166). I was hoping this was going to be merged before another release occurred ([v1.73.0 was recently released](https://github.com/JuliaCloud/AWS.jl/pull/523))

**Pull request merged**